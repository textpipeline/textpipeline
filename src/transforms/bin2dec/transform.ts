import Decimal from 'decimal.js';
import { decimalJs } from 'oss';

import { Transform } from '../types';

const containsNonBinary = (suspect: string): boolean => /[^0-1]+/.test(suspect);

const transform: Transform = {
  name: 'Binary To Decimal',
  slug: 'binary-to-decimal',
  inputTypeName: 'Binary',
  outputTypeName: 'Decimal',
  openSourceProject: decimalJs,
  defaultOutput: '0',
  execute: async input => {
    if (!input) {
      throw new Error('No input detected.');
    }

    const sanitized = input.replace(/\s+/g, '');

    if (sanitized === '') {
      throw new Error('No input detected.');
    }

    if (containsNonBinary(sanitized)) {
      throw new Error('Invalid binary input.');
    }

    const decimal = new Decimal(`0b${sanitized}`);
    return decimal.toFixed(0);
  },
};

export default transform;
