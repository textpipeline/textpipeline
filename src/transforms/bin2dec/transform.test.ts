import bin2dec from './transform';

describe('bin2dec', () => {
  it('has the correct attributes', () => {
    expect(bin2dec).toMatchSnapshot();
  });

  describe('execute', () => {
    it('rejects with a no-input error when supplied undefined', () =>
      expect(bin2dec.execute(undefined as any)).rejects.toEqual(new Error('No input detected.')));

    it('rejects with a no-input error when supplied empty', () =>
      expect(bin2dec.execute('')).rejects.toEqual(new Error('No input detected.')));

    it('rejects with a no-input error when supplied only whitespace', () =>
      expect(bin2dec.execute(' ')).rejects.toEqual(new Error('No input detected.')));

    it('rejects with an invalid-input error when supplied alpha characters', () =>
      expect(bin2dec.execute('a')).rejects.toEqual(new Error('Invalid binary input.')));

    it('rejects with an invalid-input error when supplied non-binary numeric characters', () =>
      expect(bin2dec.execute('2')).rejects.toEqual(new Error('Invalid binary input.')));

    it('resolves to the correct decimal representation when supplied a binary number', () =>
      expect(bin2dec.execute('0111')).resolves.toEqual('7'));

    it('resolves to the correct decimal representation when supplied binary with whitespace in front', () =>
      expect(bin2dec.execute(' 0111')).resolves.toEqual('7'));

    it('resolves to the correct decimal representation when supplied binary with whitespace in back', () =>
      expect(bin2dec.execute('0111 ')).resolves.toEqual('7'));

    it('resolves to the correct decimal representation when supplied a binary number with spaces', () =>
      expect(bin2dec.execute('01 11')).resolves.toEqual('7'));
  });
});
