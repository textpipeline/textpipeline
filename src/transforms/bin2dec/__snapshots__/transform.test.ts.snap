// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bin2dec has the correct attributes 1`] = `
Object {
  "defaultOutput": "0",
  "execute": [Function],
  "inputTypeName": "Binary",
  "name": "Binary To Decimal",
  "outputTypeName": "Decimal",
  "project": Object {
    "description": "An arbitrary-precision Decimal type for JavaScript.",
    "license": Object {
      "href": "https://choosealicense.com/licenses/mit",
      "name": "MIT",
      "slug": "mit",
    },
    "licenseText": "
The MIT Licence.

Copyright (c) 2019 Michael Mclaughlin

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
",
    "name": "decimal.js",
    "projectHref": "http://mikemcl.github.io/decimal.js/",
    "repositoryHref": "https://github.com/MikeMcl/decimal.js/",
    "slug": "decimal-js",
  },
  "slug": "binary-to-decimal",
}
`;
